README.md
# AtCoder-Jump-to-Submissions-from-Standings

Jump to personal submissions from standings score in AtCoder.

## Description

This user script was inspired by [Codeforces](https://codeforces.com/) standings page and [the tweet](https://twitter.com/jupijupi111/status/1220548897233231873).

***DEMO***

![](images/atcoder_jump_to_submissions_from_standings.gif)

## Features

- Show personal AC codes when the standings score was double-clicked.
  - Standings type
    - Standings
    - Extended Standings
    - Virtual Standings
    - Standings (Mult)
    - Team Standings
      - Note: Show only individual participation.
  - Relationships between scores and submissions
    - Total Score    : All of the contestant.
    - Task A, B, ... : Task A, B, ... only.

## Getting started

1. Install userscirpt manager ([Tampermonkey](https://www.tampermonkey.net/)) in your browser.
2. Install [Me](https://greasyfork.org/ja/scripts/397528-atcoder-jump-to-submissions-from-standings).
3. Click Tampermokey icon, turn on the userscript "Enabled" in dashboard.

### Usage

1. Select a standings page in AtCoder.
2. Double click on the contestant's score you want to watch submissions.
3. You can view the contestant's submissions. Enjoy!

## How to clone the repository

Paste the following commands at a Terminal prompt.

```terminal
$ mkdir hoge
$ cd hoge
$ git clone git@github.com:KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings.git
```

## Links

[AtCoder](https://atcoder.jp/)

[Original tweet](https://twitter.com/jupijupi111/status/1220548897233231873)

[User script guide](https://simply-how.com/enhance-and-fine-tune-any-web-page-the-complete-user-scripts-guide)

[User script for beginner](https://qiita.com/i_completely_understand/items/acf3e5efe0db848989d9)

Readme Driven Development; RDD<sup>[archive.org](http://web.archive.org/web/20220313000343/https://qiita.com/b4b4r07/items/c80d53db9a0fd59086ec)</sup>

## Author

[@KATO-Hiro](https://twitter.com/k_hiro1818)

## License

[CC0](https://creativecommons.org/share-your-work/public-domain/cc0)


CHANGELOG.md
# Change Log

## 0.2 / 2022-10-23

- [#66](https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/pull/66) - ヒューリスティックコンテストの「延長戦順位表」ページから、「提出結果」ページが見られるように修正。依存ライブラリをアップデート。

## 0.1.1 / 2020-10-14

- [#58](https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/pull/58), [#59](https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/pull/59) - ライセンスをMITからCC0に変更。
- [#57](https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/pull/57) - READMEに「チーム戦順位表」への対応と注意事項を追記。
- [#54](https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/pull/54) - コメントアウトした古いコードを削除。
- [#52](https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/pull/52) - 「チーム戦順位表」ページから、「提出結果」ページが見られるように修正。


AtCoder Jump to Submissions from Standings.user.js
// ==UserScript==
// @name         AtCoder Jump to Submissions from Standings
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  順位表の得点をダブルクリックすると、該当するコンテスタントの実装を見ることができます。
// @match        https://atcoder.jp/contests/*/standings*
// @require      https://code.jquery.com/jquery-3.4.1.min.js
// @author       hiro_hiro
// @license      MIT
// @downloadURL  https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/atcoder_jump_to_submissions_from_standings.js
// @updateURL    https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/atcoder_jump_to_submissions_from_standings.js
// @supportURL   https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/issues
// @grant        none
// ==/UserScript==

$(function() {
    'use strict';

    $(document).on('dblclick', '.standings-result', function() {
        const $standingsType = getStandingsType($('body'));

        const $prefix = addPrefixIfNeeds($standingsType);

        const $clickedColumnIndex = getClickedColumnIndex(this, $standingsType);
        const $taskUrls = $('body').find('thead a');
        const $taskId = getTaskId($taskUrls, $clickedColumnIndex);

        const $username = getUserName(this);

        const $displayLanguage = getDisplayLanguage($(location));
        const $suffix = addSuffixIfNeeds($displayLanguage);

        // 順位表の範囲外なら、提出ページに遷移しない
        if ($clickedColumnIndex < $taskUrls.length) {
            // jumpToPersonalSubmissions($prefix, $taskId, $username);
            jumpToPersonalSubmissions($prefix, $taskId, $username, $suffix);
        }
    });
})();

function getStandingsType(object) {
    let $standingsType = '';
    let isVirtual = $(object).find('script:contains("virtual")')[0];
    let isMultiply = $(object).find('script:contains("multiply_ranks")')[0];

    // HACK: if分岐はメンテナンス的によくないかも
    // HACK: 他の言語のEnumに相当する構文がデフォルトで存在しない?
    if (isVirtual) {
        $standingsType = 'virtual';
    } else if (isMultiply) {
        $standingsType = 'multiply';
    } else {
        $standingsType = 'general';
    }

    return $standingsType
}

function addPrefixIfNeeds(standingsType) {
    let prefix = '';

    if (standingsType != 'general') {
        prefix = '../';
    }

    return prefix
}

function getTaskId(taskUrls, clickedColumnIndex) {
    let $taskId = '';

    taskUrls.each((index) => {
        if (index == clickedColumnIndex) {
            const $url = taskUrls[index].pathname;
            const $elements = $url.split('/');
            const $length = $elements.length;

            $taskId = $elements[$length - 1]; // 0-indexed
        }
    });

    return $taskId
}

// HACK: 順位表の列数に応じた処理をしているため、AtCoderのUIが変更されると動かなくなる可能性がある
// WHY : 順位表の得点の欄に、問題のIDが含まれていないため
function getClickedColumnIndex(object, standingsType) {
    let $clickedColumnIndex = $(object)[0].cellIndex;

    // コンテスト当日の順位表とバーチャル順位表の列の並びに違いがある
    // 当日の順位表の並びに合わせる
    if (standingsType == 'virtual') {
        $clickedColumnIndex -= 1
    }

    // 順位とユーザ名の欄を扱わずに済むようにインデックスを補正
    // A問題がindex = 0となるようにしている
    $clickedColumnIndex -= 3

    return $clickedColumnIndex
}

function getUserName(object) {
    const $standings = $(object).siblings('td');
    const $username = $standings.find('.username span').text();

    return $username
}

function getDisplayLanguage(location) {
    let language = 'jp';
    const params = location.attr('search');

    if (params.match(/lang=en/)) {
        language = 'en';
    }

   return language
}

function addSuffixIfNeeds($displayLanguage) {
    let suffix = '&lang=';

    if ($displayLanguage == 'en') {
        suffix += 'en';
    } else {
        suffix = '';
    }

    return suffix
}

function jumpToPersonalSubmissions(prefix, taskId, username, suffix) {
    setTimeout(function() {
        location.href = `${prefix}submissions?f.Task=${taskId}&f.Language=&f.Status=AC&f.User=${username}${suffix}`;
    }, 250)
}


AtCoder_Jump_to_Submissions_from_Standings.user.js
// ==UserScript==
// @name         AtCoder Jump to Submissions from Standings
// @namespace    http://tampermonkey.net/
// @version      0.2
// @description  順位表の得点をダブルクリックすると、該当するコンテスタントの実装を見ることができます。
// @match        https://atcoder.jp/contests/*/standings*
// @require      https://code.jquery.com/jquery-3.6.1.min.js
// @author       hiro_hiro
// @license      CC0
// @downloadURL  https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/AtCoder_Jump_to_Submissions_from_Standings.user.js
// @updateURL    https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/AtCoder_Jump_to_Submissions_from_Standings.user.js
// @supportURL   https://github.com/KATO-Hiro/AtCoder-Jump-to-Submissions-from-Standings/issues
// @grant        none
// ==/UserScript==

$(function () {
    'use strict';

    $(document).on('dblclick', '.standings-result', function () {
        const $standingsType = getStandingsType($('body'));

        const $prefix = addPrefixIfNeeds($standingsType);

        const $clickedColumnIndex = getClickedColumnIndex(this, $standingsType);
        const $taskUrls = $('body').find('thead a');
        const $taskId = getTaskId($taskUrls, $clickedColumnIndex);

        const $username = getUserName(this);

        const $displayLanguage = getDisplayLanguage($(location));
        const $suffix = addSuffixIfNeeds($displayLanguage);

        // 順位表の範囲外なら、提出ページに遷移しない
        if ($clickedColumnIndex < $taskUrls.length) {
            jumpToPersonalSubmissions($prefix, $taskId, $username, $suffix);
        }
    });
})();

function getStandingsType(object) {
    let $standingsType = '';
    let isVirtual = $(object).find('script:contains("virtual")')[0];
    let isMultiply = $(object).find('script:contains("multiply_ranks")')[0];
    let isTeam = $(object).find('script:contains("team")')[0];
    let isExtended = $(object).find('script:contains("extended")')[0];

    // HACK: if分岐はメンテナンス的によくないかも
    // HACK: 他の言語のEnumに相当する構文がデフォルトで存在しない?
    if (isVirtual) {
        $standingsType = 'virtual';
    } else if (isMultiply) {
        $standingsType = 'multiply';
    } else if (isTeam) {
        $standingsType = 'team';
    } else if (isExtended) {
        $standingsType = 'extended';
    } else {
        $standingsType = 'general';
    }

    return $standingsType
}

function addPrefixIfNeeds(standingsType) {
    let prefix = '';

    if (standingsType != 'general') {
        prefix = '../';
    }

    return prefix
}

function getTaskId(taskUrls, clickedColumnIndex) {
    let $taskId = '';

    taskUrls.each((index) => {
        if (index == clickedColumnIndex) {
            const $url = taskUrls[index].pathname;
            const $elements = $url.split('/');
            const $length = $elements.length;

            $taskId = $elements[$length - 1]; // 0-indexed
        }
    });

    return $taskId
}

// HACK: 順位表の列数に応じた処理をしているため、AtCoderのUIが変更されると動かなくなる可能性がある
// WHY : 順位表の得点の欄に、問題のIDが含まれていないため
function getClickedColumnIndex(object, standingsType) {
    let $clickedColumnIndex = $(object)[0].cellIndex;

    // コンテスト当日の順位表と延長戦順位表・バーチャル順位表の列の並びに違いがある
    // 当日の順位表の並びに合わせる
    if (standingsType == 'virtual' || standingsType == 'extended') {
        $clickedColumnIndex -= 1
    }

    // 順位とユーザ名の欄を扱わずに済むようにインデックスを補正
    // A問題がindex = 0となるようにしている
    $clickedColumnIndex -= 3

    return $clickedColumnIndex
}

function getUserName(object) {
    const $standings = $(object).siblings('td');
    const $username = $standings.find('.username span').text();

    return $username
}

function getDisplayLanguage(location) {
    let language = 'jp';
    const params = location.attr('search');

    if (params.match(/lang=en/)) {
        language = 'en';
    }

    return language
}

function addSuffixIfNeeds($displayLanguage) {
    let suffix = '&lang=';

    if ($displayLanguage == 'en') {
        suffix += 'en';
    } else {
        suffix = '';
    }

    return suffix
}

function jumpToPersonalSubmissions(prefix, taskId, username, suffix) {
    setTimeout(function () {
        location.href = `${prefix}submissions?f.Task=${taskId}&f.Language=&f.Status=AC&f.User=${username}${suffix}`;
    }, 250)
}
